packages/report-server/src/__tests__/reportBuilder/transform/parser/functions.test.ts

```diff
@@ -0,0 +1,172 @@
+    });
+
+    describe('periodToTimestamp', () => {
+      it('converts given period to timestamp', () =>
+        expect(new TransformParser().evaluate("=periodToTimestamp('20200101')")).toBe(
+          1577836800000,
+        ));
+    });
+
+    describe('periodToDisplayString', () => {
+      it('converts given period to display string of type', () =>
+        expect(new TransformParser().evaluate("=periodToDisplayString('20200101', 'WEEK')")).toBe(
+          '1st Jan 2020',
+        ));
+    });
+
+    describe('dateStringToPeriod', () => {
+      it('converts given date string to period', () =>
+        expect(new TransformParser().evaluate("=dateStringToPeriod('2020-02-15', 'WEEK')")).toBe(
+          '2020W07',
+        ));
+    });
+
+    describe('formatAsFractionAndPercentage', () => {
+      it('formats values as fraction and percentage', () =>
+        expect(new TransformParser().evaluate('=formatAsFractionAndPercentage(1, 10)')).toBe(
+          '1/10 = 10%',
+        ));
+
+      it('throws error if denominator is 0', () =>
+        expect(() => {
+          new TransformParser().evaluate('=formatAsFractionAndPercentage(1, 0)');
+        }).toThrow());
+    });
+  });
+
+  describe('math', () => {
+    describe('add', () => {
+      it('returns first number if second is undefined', () =>
+        expect(new TransformParser().evaluate('=1 + undefined')).toBe(1));
+
+      it('returns second number if first is undefined', () =>
+        expect(new TransformParser().evaluate('=undefined + 5')).toBe(5));
+
+      it('returns undefined if all values are undefined', () =>
+        expect(new TransformParser().evaluate('=undefined + undefined')).toBe(undefined));
+    });
+
+    describe('divide', () => {
+      it('returns undefined if second number is undefined', () =>
+        expect(new TransformParser().evaluate('=1 / undefined')).toBe(undefined));
+
+      it('returns undefined if first number is undefined', () =>
+        expect(new TransformParser().evaluate('=undefined / 5')).toBe(undefined));
+
+      it('returns undefined if both numbers are undefined', () =>
+        expect(new TransformParser().evaluate('=undefined / undefined')).toBe(undefined));
+    });
+
+    describe('sum', () => {
+      it('throws error for non-numeric input', () =>
+        expect(() => new TransformParser().evaluate("=sum('cat')")).toThrow());
+
+      it('throws error for non-numeric array input', () =>
+        expect(() => new TransformParser().evaluate("=sum([1, 'cat'])")).toThrow());
+
+      it('throws error for non-numeric inputs', () =>
+        expect(() => new TransformParser().evaluate("=sum(1, 'cat')")).toThrow());
+
+      it('returns undefined if input is undefined', () =>
+        expect(new TransformParser().evaluate('=sum(undefined)')).toBe(undefined));
+
+      it('returns undefined if input is array of undefined', () =>
+        expect(new TransformParser().evaluate('=sum([undefined, undefined])')).toBe(undefined));
```

Comment:
Suggestion: it may be a bit subjective but for repeated test cases I often find [it.each](https://jestjs.io/docs/api#testeachtablename-fn-timeout) very helpful, and easier to maintain. Most of all, it encourages you to think of more edge cases, since you can add them very easily!

We have a lot of examples in our code base, eg in [array.test.js](https://github.com/beyondessential/tupaia/blob/dev/packages/utils/src/__tests__/array.test.js#L74). This section describe block would look something like the following:

```ts
describe('sum', () => {
  const testData: [string, string, unknown][] = [
    ['returns undefined if input is undefined', '=sum(undefined)', undefined],
    ['returns undefined if input is array of undefined', '=sum([undefined, undefined])', undefined],
    ['returns undefined if inputs are undefined', '=sum(undefined, undefined)', undefined],
    ['returns sum if input is at least partially defined array', 1],
    // ...etc
  ];

  it.each(testData)('%s', (_, expression: string, expected: unknown) => {
    expect(new TransformParser().evaluate(expression)).toBe(expected);
  });
});

```

Reply:
On the other hand, your test cases here are already easy to read and elegant, so I wouldn't necessarily "recommend" using `it.each` here. I'd probably do it because I like the concept of [parameterized tests](https://www.baeldung.com/parameterized-tests-junit-5), but that's probably a matter of taste.

Reply:
I agree, it's easy to go overboard with parameterized tests!_